chanfitness/chaneventstore.go

```diff
@@ -0,0 +1,308 @@
+			case <-c.quit:
+			case <-channelClient.Quit():
+			case <-peerClient.Quit():
+				close(events)
+				return
+			}
+		}
+	}()
+
+	// Add the existing set of channels to the event store. This is required
+	// because channel events will not be triggered for channels that exist
+	// at startup time.
+	channels, err := c.cfg.GetOpenChannels()
+	if err != nil {
+		return err
+	}
+
+	log.Infof("Adding %v channels to event store", len(channels))
+
+	for _, ch := range channels {
+		var peerKey [33]byte
+		copy(peerKey[:], ch.IdentityPub.SerializeCompressed())
+
+		// Add existing channels to the channel store with an initial peer
+		// online or offline event.
+		c.addChannel(ch.ShortChanID().ToUint64(), peerKey)
+	}
+
+	return nil
+}
+
+// Stop terminates all goroutines started by the event store.
+func (c *ChannelEventStore) Stop() {
+	log.Info("Stopping event store")
+
+	// Stop the monitorChannelEvents goroutine.
+	close(c.quit)
+
+	c.wg.Wait()
+}
+
+// addChannel adds a new channel to the ChannelEventStore's map of channels with
+// an initial peer online state (if the peer is online). The channel's open time
+// is set after the online event so that we do not record a short offline period
+// in between online and open time. If the channel is already present in the map,
+// the function returns early. This function should be called to add existing
+// channels on startup and when open channel events are observed.
+func (c *ChannelEventStore) addChannel(channelID uint64, peer [33]byte) {
+	c.Lock()
+	defer c.Unlock()
+
+	_, ok := c.channels[channelID]
+	if ok {
+		return
+	}
+
+	eventLog := &chanEventLog{
+		id:   channelID,
+		peer: peer,
+		now:  time.Now,
+	}
+
+	// If the peer is online, add a peer online event to indicate its starting
+	// state.
+	online := c.peers[peer]
+	if online {
+		eventLog.add(peerOnlineEvent)
+	}
+
+	// Set the open time of the eventLog after the initial online event so that
+	// we do not incorrectly record a short offline period when the peer is
+	// actually online.
+	eventLog.openedAt = eventLog.now()
+
+	c.channels[channelID] = eventLog
+}
+
+// closeChannel records a closed time for a channel, and returns early is the
+// channel is not known to the event store.
+func (c *ChannelEventStore) closeChannel(channelID uint64) {
+	c.Lock()
+	defer c.Unlock()
+
+	eventLog, ok := c.channels[channelID]
+	if !ok {
+		// Return early the channel is not known, since we have not been
+		// tracking events for it anyway.
+		return
+	}
+
+	eventLog.closedAt = eventLog.now()
+}
+
+// peerEvent adds a peer online or offline event to all channels we currently
+// have open with a peer.
+func (c *ChannelEventStore) peerEvent(peer [33]byte, event eventType) {
+	c.Lock()
+	defer c.Unlock()
+
+	// Track current online status of peers in the channelEventStore.
+	peerIsOnline := true
+	if event == peerOfflineEvent {
+		peerIsOnline = false
+	}
+	c.peers[peer] = peerIsOnline
```

Comment:
is the assumption that we will start listening for these peer events before lnd connects to any peers?

Reply:
Yes. If I'm following properly, all subservers are kicked off in `server.go` before we connect to peers `server.go#1333`.

This does create an issue where channels that we have open on startup will record their peers being offline for the period that it takes to connect. I thought about adding some kind of sleep but I think the various impls have quite different connection times?

Reply:
Sounds good, I think this is good enough 👍